src/passes/MultiMemoryLowering.cpp

```diff
@@ -0,0 +1,394 @@
+      assert(memory->shared == isShared);
+      assert(memory->indexType == pointerType);
+
+      // Calculating the total initial and max page size for the combined memory
+      // by totaling the initial and max page sizes for the memories in the
+      // module
+      totalInitialPages = totalInitialPages + memory->initial;
+      if (memory->hasMax()) {
+        totalMaxPages = totalMaxPages + memory->max;
+      }
+    }
+    // Ensuring valid initial and max page sizes that do not exceed the number
+    // of pages addressable by the pointerType
+    Address maxSize =
+      pointerType == Type::i32 ? Memory::kMaxSize32 : Memory::kMaxSize64;
+    if (totalMaxPages > maxSize || totalMaxPages == 0) {
+      totalMaxPages = Memory::kUnlimitedSize;
+    }
+    if (totalInitialPages > totalMaxPages) {
+      totalInitialPages = totalMaxPages;
+    }
+
+    // Creating the combined memory name so we can reference the combined memory
+    // in subsequent instructions before it is added to the module
+    combinedMemory = Names::getValidMemoryName(*wasm, "combined_memory");
+  }
+
+  void addOffsetGlobals() {
+    auto addGlobal = [&](Name name, size_t offset) {
+      auto global =
+        Builder::makeGlobal(name,
+                            pointerType,
+                            Builder(*wasm).makeConst(
+                              pointerType == Type::i32
+                                ? Literal::makeFromInt32(offset, pointerType)
+                                : Literal::makeFromInt64(offset, pointerType)),
+                            Builder::Mutable);
+      global->hasExplicitName = true;
+      wasm->addGlobal(std::move(global));
+    };
+
+    size_t offsetRunningTotal = 0;
+    for (Index i = 0; i < wasm->memories.size(); i++) {
+      auto& memory = wasm->memories[i];
+      memoryIdxMap[memory->name] = i;
+      // We don't need a page offset global for the first memory as it's always
+      // 0
+      if (i != 0) {
+        Name name = Names::getValidGlobalName(
+          *wasm, memory->name.toString() + "_page_offset");
+        offsetGlobalNames.push_back(std::move(name));
+        addGlobal(name, offsetRunningTotal);
+      }
+      offsetRunningTotal += memory->initial;
+    }
+  }
+
+  void adjustActiveDataSegmentOffsets() {
+    Builder builder(*wasm);
+    ModuleUtils::iterActiveDataSegments(*wasm, [&](DataSegment* dataSegment) {
+      assert(dataSegment->offset->is<Const>() &&
+             "TODO: handle non-const segment offsets");
+      auto idx = memoryIdxMap.at(dataSegment->memory);
+      dataSegment->memory = combinedMemory;
+      // No need to update the offset of data segments for the first memory
+      if (idx != 0) {
+        // We need to subtract 1 here because there is no offsetGlobal for the
+        // first memory
+        auto offsetGlobalName = getOffsetGlobal(idx);
+        dataSegment->offset = builder.makeBinary(
+          Abstract::getBinary(pointerType, Abstract::Add),
```

Comment:
I believe an add in a data segment offset requires the ExtendedConst feature, so we should error if `!wasm->features.hasExtendedConst()`